res/res_pjsip_session.c

```diff
@@ -5401,6 +5611,8 @@ static void session_inv_on_create_offer(pjsip_inv_session *inv, pjmedia_sdp_sess
 		SCOPE_EXIT_RTN("%s: create offer failed\n", ast_sip_session_get_name(session));
 	}
 
+	ast_sip_session_media_state_reset(session->pending_media_state);
```

Comment:
This seems to cause no audio with `all_codecs_on_empty_reinvite=yes` when a UA renegotiates a different codec. See [1] for reproducing this issue. Seems to be fixed with
```
diff --git a/res/res_pjsip_session.c b/res/res_pjsip_session.c
index 5486a97040..72b4130a40 100644
--- a/res/res_pjsip_session.c
+++ b/res/res_pjsip_session.c
@@ -5594,25 +5594,23 @@ static void session_inv_on_create_offer(pjsip_inv_session *inv, pjmedia_sdp_sess
 		}
 	}
 
-	if (inv->neg) {
-		if (pjmedia_sdp_neg_was_answer_remote(inv->neg)) {
-			pjmedia_sdp_neg_get_active_remote(inv->neg, &previous_sdp);
-		} else {
-			pjmedia_sdp_neg_get_active_local(inv->neg, &previous_sdp);
-		}
-	}
-
 	if (ignore_active_stream_topology) {
 		offer = create_local_sdp(inv, session, NULL, 1);
 	} else {
+		if (inv->neg) {
+			if (pjmedia_sdp_neg_was_answer_remote(inv->neg)) {
+				pjmedia_sdp_neg_get_active_remote(inv->neg, &previous_sdp);
+			} else {
+				pjmedia_sdp_neg_get_active_local(inv->neg, &previous_sdp);
+			}
+		}
 		offer = create_local_sdp(inv, session, previous_sdp, 0);
+		ast_sip_session_media_state_reset(session->pending_media_state);
 	}
 	if (!offer) {
 		SCOPE_EXIT_RTN("%s: create offer failed\n", ast_sip_session_get_name(session));
 	}
 
-	ast_sip_session_media_state_reset(session->pending_media_state);
-
 	ast_queue_unhold(session->channel);
 
 	/*
```

[1] [acn_reinvite_without_sdp.patch](https://github.com/asterisk/asterisk/files/13270602/acn_reinvite_without_sdp.patch)

Reply:
I don't doubt it, but the provided testcase passes for me even without your suggested patch. Is it testing exactly the scenario you described?

Reply:
Negotiating codecs works fine without the patch but there's no audio (no RTP sent) after negotiating a new codec. I don't know how to verify that RTP is sent (maybe it's possible to check for RTP AMI events?). IIRC without the patch there's following log message
```
res_pjsip_sdp_rtp.c:487 set_caps: No joint capabilities for 'audio' media stream between our configuration((alaw)) and incoming SDP((g722))
```

Reply:
Makes sense, maybe I can find a way to test it. Thank you for clarifying.